### ISSUE TITLE
Warn if query is executed multiple times

### ISSUE THREAD

**Issue by Hyperblaster** [Automattic/mongoose#5683]
I have read through multiple issues and discussions here but can not find a definitive workaround for this requirement as 

1. Asynchronous setters are not possible
2. pre('validate') hook does not fire until after the schema validation

What i am needing to do is something like this _(code reduced for brevity)_

**My Schema**
`
var MySchema = new Schema({
title:{
 type:String,
 required:true,
},
tags: [{
        type: ObjectId,
        ref: 'tag',
    }]
});
`

**Data i am trying to save**
`
var myDocument = new MySchemaDocument({
  "title":"something",
  "tags":["Tag Title 1", "Tag Title 2"],
})
`

**My validate hook function** 
_(Searches data base and convert submitted strings to matching object ids)_
`
MySchema.pre('validate', function(next) {
  var doc = this;
  doc.tags = _.map(doc.tags, function(tag) {
       return 'MYOBJECTID'
       //Remap the tags to object ids and then call next()
  })
});
`

**What happens?**
The mongoose validation fails due to the array of strings being sent through not being valid ObjectIds, obviously i want my PRE validate middleware to convert the strings into objectIDs, but it seems that the pre('validate') hook does not actually run before the schema validates but after it.

reading through other discussions, possible workarounds were to use the pre('set') hook and try and adjust it there, (doesn't work) or to use a custom setter, (cant solve this issue because i need it to be asynchronous.

Any other thoughts for workarounds here?

**Comment by vkarpov15** [comment]
Below script works fine for me in 4.12.0, no errors reported. Can you modify the below script to reproduce your issue?

```javascript
const mongoose = require('mongoose')

var MySchema = new mongoose.Schema({
  title:{ type:String, required:true },
  tags: [{ type: mongoose.Schema.Types.ObjectId, ref: 'tag' }]
});

MySchema.pre('validate', function(next) {
  this.tags = ['000000000000000000000000'];
  next();
});

var MyModel = mongoose.model('MyModel', MySchema);

var doc = new MyModel({ title: 'test', tags: ['bad'] });

doc.validate(function(error) {
  console.log('Error', error);
});
```

**Comment by Hyperblaster** [comment]
I'm using 4.11.14 i'll update and check it out

**Comment by PigBoT** [comment]
@Hyperblaster Did you get anywhere with this?  I may be having an issue as well, using 4.12.2.  I'm debugging now.

**Comment by vkarpov15** [comment]
@PigBoT please provide code samples :+1: :palm_tree:

**Comment by felipedonda** [comment]
@Hyperblaster I was caught in the same situation but I understood what happened: The pre validation hook ended before your could do your conversion because you were doing it inside a callback, which makes Node understand that there's nothing else left to wait and it's free to end the hook.

**Comment by felipedonda** [comment]
@Hyperblaster I was caught in the same situation but I understood what happened: The pre validation hook ended before your could do your conversion because you were doing it inside a callback but you were calling `next()` outside of it. Node will not wait your callback to end before calling 'next()', you need to call it inside your callback to make sure the order of execution is right.

Instead of this:
`Schema.pre('validate', function(next) {
    function( callback(){
        //stuff
    })
    next()
}) `

Do this:
`Schema.pre('validate', function(next) {
    function( callback(){
        //stuff
        next()
    })
}) `

**Comment by jwerre** [comment]
Here is a code sample regarding this issue. This will fail *some*times. It fails is because a Query object is returned inside the pre hook.

```
const assert = require('assert'),
	mongoose = require('mongoose');

const bookSchema = new mongoose.Schema({
	index: {
		type: Number,
		required: true
	},
	title: {
		type: String,
	},
	author: {
		type: String,
	},
	genre: {
		type: String,
		enum: ['fiction', 'non-fiction'],
		default: 'fiction'
	},
	created: {
		type: Date,
		default: Date.now
	},
	updated: {
		type: Date,
		default: Date.now
	},
}, {
	timestamps: {
		createdAt: 'created',
		updatedAt: 'updated'
	}
});

bookSchema.index({ author: 1 });

bookSchema.pre('validate', function(callback) {

	console.log('Book pre::validate()');

	if (this.index != null) {
		return callback();
	}	

	let _this = this;
	
	// Book.countDocuments( // <--- this does not fail
	return Book.countDocuments( // <--- this will fail sometimes

		{
			author: this.author
		}, 
		
		function(err, count) {

			if (err || !count) {
				count = 0;
			}
			
			_this.index = count;

			console.log( 'Setting Index:', _this.index);
			
			callback();
		}
	);

});

const Book = mongoose.model('Book', bookSchema);

describe('Book', function() {
	
	let bookId;
	
	before(function (done) {
		
		mongoose.connect( 'mongodb://localhost/test', {
			poolSize: 30,
			useNewUrlParser: true,
			useCreateIndex: true,
			useFindAndModify: false,
			autoIndex: false,
			reconnectTries: Number.MAX_VALUE,
			reconnectInterval: 500,
			keepAlive: true,
			keepAliveInitialDelay: 300000,
			connectTimeoutMS: 10000,
			socketTimeoutMS: 45000,
			family: 4
		}, done );
			
	});

	after(function (done) {
		mongoose.disconnect(done);
	});


	it('should save a new book', function(done) {
		
		var myModel = new Book();
		myModel.title = 'Ulysses';
		myModel.author = 'James Joyce';
		myModel.created = new Date();
		
		myModel.save( function(err, doc) {
			assert.ifError(err);
			assert.ok(doc);
			bookId = doc._id;
			done(err);
		});

	
	});
	
	it('should get the new book', function(done) {
	
		Book.findById( bookId, function(err, doc) {
			assert.ifError(err);
			assert.ok(doc);
			return done(err);
		});
	
	});


	it('should save a new book', function(done) {
		
		var myModel = new Book();
		myModel.title = 'Crime and Punishment';
		myModel.author = 'Foyder Dostoevsky';
		myModel.created = new Date();
		
		myModel.save( function(err, doc) {
			assert.ifError(err);
			assert.ok(doc);
			bookId = doc._id;
			done(err);
		});

	
	});
	
	it('should get the new book', function(done) {
	
		Book.findById( bookId, function(err, doc) {
			assert.ifError(err);
			assert.ok(doc);
			return done(err);
		});
	
	});
	
});
```

Again, this might not fail but if you run it a bunch of time it will. Try something like this:

    for run in {1..10}; do mocha ./test.js; done


Versions:
node: v8.12.0
mongoose: v5.4.0
mongoDB: 4.0.4

**Comment by vkarpov15** [comment]
@jwerre I looked closer into this and the issue is that your `countDocuments()` query is executing twice, once because you passed a callback and once because you're returning a thenable from middleware. You then have a race condition between the two because the execution from the thenable doesn't call the callback. This is a bit of a gotcha, I opened up #7398 to track and will make some improvements to make catching this easier in the future.
